### PR TITLE
Fix issue 17333 for except over no columns.

### PR DESCRIPTION
**Pull request by z-wenlin**
This PR is trying to fix the following problem.
postgres=# create table tt2(id int);
postgres=# insert into tt2 select generate_series(1,5); postgres=# set optimizer=off;
SET
postgres=# explain select from generate_series(1,10) except select from tt2; 
ERROR:  unexpected gang size: 3 (nodeMotion.c:732)

In the adjust_setop_arguments(), if the setop is partitioned, the make_motion_hash_all_targets() will change the locus to singleQE when none of the columns are hashable. But mark_append_locus() will change it back to the original partitioned locus, adding a Motion node at the top. It will generate the following plan
```
 postgres=#  explain select from generate_series(1,10) INTERSECT select from tt2;
                                             QUERY PLAN

-----------------------------------------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)  (cost=0.00..2441.77 rows=1 width=4)
   ->  HashSetOp Intersect  (cost=0.00..2441.75 rows=1 width=4)
         ->  Append  (cost=0.00..2441.75 rows=96310 width=4)
               ->  Subquery Scan on "*SELECT* 1"  (cost=0.00..0.20 rows=10 width=4)
                     ->  Function Scan on generate_series  (cost=0.00..0.10 rows=10 width=0)
               ->  Gather Motion 3:1  (slice2; segments: 3)  (cost=0.00..1960.00 rows=96300 width=4)
                     ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..676.00 rows=32100 width=4)
                           ->  Seq Scan on tt2  (cost=0.00..355.00 rows=32100 width=0)
 Optimizer: Postgres-based planner
(9 rows) 
```
When ExecInitMotion() for the second motion, the recvSlice->segments is not equal to 1, which results in the error.

This PR moves the check to choose_setop_type() when there is no target list and returns the singleQE locus. 

The dev pipeline is https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/wz.big_lag

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

File: src/backend/cdb/cdbsetop.c [hunk 1]
```diff
@@ -320,7 +320,13 @@ mark_append_locus(Path *path, GpSetOpType optype)
 			CdbPathLocus_MakeGeneral(&path->locus);
 			break;
 		case PSETOP_PARALLEL_PARTITIONED:
-			CdbPathLocus_MakeStrewn(&path->locus, getgpsegmentCount());
+			if ((&path->locus)->locustype == CdbLocusType_SingleQE )
```
Comment (avamingli) [inline src/backend/cdb/cdbsetop.c:323]: use `CdbPathLocus_IsSingleQE()`?

**Comment by z-wenlin** [comment]
The dev pipeline is https://dev.ci.gpdb.pivotal.io/teams/main/pipelines/wz.big_lag

File: src/backend/cdb/cdbsetop.c [hunk 2]
```diff
@@ -320,7 +320,13 @@ mark_append_locus(Path *path, GpSetOpType optype)
 			CdbPathLocus_MakeGeneral(&path->locus);
 			break;
 		case PSETOP_PARALLEL_PARTITIONED:
-			CdbPathLocus_MakeStrewn(&path->locus, getgpsegmentCount());
+			if (CdbPathLocus_IsSingleQE(path->locus))
```
Comment (charliettxx) [inline src/backend/cdb/cdbsetop.c:323]: Is that possible mark optype PSETOP_SEQUENTIAL_QE if none of key are hashable?

File: src/backend/cdb/cdbsetop.c [hunk 3]
```diff
@@ -39,19 +39,27 @@
...
 	bool		has_partitioned = false;
 
 	Assert(Gp_role == GP_ROLE_DISPATCH || Gp_role == GP_ROLE_UTILITY);
 
-	foreach(cell, pathlist)
+	forboth(cell, pathlist, tlistcell, tlist_list)
 	{
 		Path	   *subpath = (Path *) lfirst(cell);
+		List	   *subtlist = (List *) lfirst(tlistcell);
+		foreach(cell, subtlist)
```
Comment (charliettxx) [inline src/backend/cdb/cdbsetop.c:57]: I suppose it should be if all of targetentry in subtlist are resjunk then return PSETOP_SEQUENTIAL_QE, not
one of them.

File: src/backend/cdb/cdbsetop.c [hunk 4]
```diff
@@ -39,19 +39,27 @@
...
+		{
+			TargetEntry *tle = (TargetEntry *) lfirst(cell);
+			if (tle->resjunk)
+				return PSETOP_SEQUENTIAL_QE;
```
Comment (charliettxx) [inline src/backend/cdb/cdbsetop.c:61]: maybe PSETOP_SEQUENTIAL_QD is better?

File: src/backend/cdb/cdbsetop.c (hunk 3 above)
Comment (z-wenlin) [reply src/backend/cdb/cdbsetop.c]: Yes, you are correct, I have changed this part.

**Review by charliettxx — APPROVED** [review]

**Pull request merged**